[Hub Generated] Review request for Microsoft.DocumentDB to add version preview/2023-11-15-preview

pjohari-ms opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/b321e1dc-5e07-4940-9cd6-a65ac6729a8b)

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

kshittiz7 committed:
swagger changes with canUndelete and canUndeleteReason flags

adityasa committed:
Added Aditya's compute property

pjohari-ms committed:
Adds base for updating Microsoft.DocumentDB from version preview/2023-09-15-preview to version 2023-11-15-preview

pjohari-ms committed:
Updates readme

pjohari-ms committed:
Updates API version in new specs and examples

niteshvijay-ms committed:
Fix DTS source sink model

niteshvijay-ms committed:
Add job mode in DTS

SrinikhilReddy committed:
Add new resource - ThroughputPool

niteshvijay-ms committed:
Add duration field in job properties

AdrianSibajaRetana committed:
Merge with Adrian's fix

pjohari-ms committed:
Corrected ThroughputPool swagger

pjohari-ms committed:
Nit

SrinikhilReddy committed:
Add ThroughputPoolAccountCreateParameters

pjohari-ms commented:
@rkmanda Gentle ping here as well. Here, we're introducing a new resource to our service.

DrewKaufmann committed:
Update MI Cassandra preview swagger (#26360)

* Update MI Cassandra preview swagger

* Fixed spellcheck and modelviolation checks

* Made model changes

* Changed param field names and added ListCommand op

* Made minor fixes

* Fixed ModelValidation checks

* Readded ARMProxyResource to BackupResource

* Removed ARMProxyResource from BackupResource

raosuhas commented:
>           "type": "string"

Looks like this PR also has the same breaking change issue as your other PR

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/cosmos-db.json:8963 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>                     "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/ApiVersionParameter"

please use v5 types for all new definitions

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:48 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>                 "nextLinkName": null

why is this set to null ?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:63 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>                 "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/SubscriptionIdParameter"

v5 for all definitions in this new file

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:78 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>                 "$ref": "#/definitions/ThroughputPoolCreateUpdateParameters"

the PUT resource and the GET resource must refer to the same type . 

You can use readonly \ x-ms-mutability to specify behavior for individual properties to differentiate between  creates \gets and updates

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:162 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>               "$ref": "#/definitions/ThroughputPoolAccountCreateParameters"

same comment about GET\PUT parity

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:281 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>             "x-ms-long-running-operation": true,

please add final-strate-via for these annotations : 

[https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:259 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>               "description": "Lists all the locations for throughput pool.",

do you not have individual gets for each location ?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:383 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>         "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/throughputPools/{poolName}/locations": {

is this designed for locationbased routing ? you do not have to strictly expose this api if you dont need to 

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:375 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>             "description": "Lists all the Azure Cosmos DB database accounts available under the subscription.",

update descriptions

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:419 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>             "readOnly": true

do you really need this ? you cant really patch the only property this has wihch is readonly right ?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:501 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>               "type": "string"

this is a bt confusing , can the ThroughputPoolAccount have a different location then the global database account ? How does this work for data residency requierments if these are different ?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:534 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

raosuhas commented:
>               "type": "string"

also why do you need to expose this location as a readonly propery here ?

---
In reply to: [1821851535](https://github.com/Azure/azure-rest-api-specs/pull/26206#issuecomment-1821851535) [](http://example.com/codeflow?ancestors=1821851535)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:534 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

pjohari-ms committed:
Addressed review comments

pjohari-ms commented:
> > ```
> >         "x-ms-long-running-operation": true,
> > ```
> 
> please add final-strate-via for these annotations :
> 
> https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:259 in [90782ad](https://github.com/Azure/azure-rest-api-specs/commit/90782addcc11916078db97f65b1e36fc09ec4b04). [](commit_id = [90782ad](https://github.com/Azure/azure-rest-api-specs/commit/90782addcc11916078db97f65b1e36fc09ec4b04), deletion_comment = False)

Not sure if we need this? According to above link it mentions that we don't need this if we follow ARM guidelines, which we do by modeling it around azure-async-operation.

pjohari-ms commented:
> > ```
> >           "type": "string"
> > ```
> 
> this is a bt confusing , can the ThroughputPoolAccount have a different location then the global database account ? How does this work for data residency requierments if these are different ?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:534 in [90782ad](https://github.com/Azure/azure-rest-api-specs/commit/90782addcc11916078db97f65b1e36fc09ec4b04). [](commit_id = [90782ad](https://github.com/Azure/azure-rest-api-specs/commit/90782addcc11916078db97f65b1e36fc09ec4b04), deletion_comment = False)

ThroughputPool by itself doesn't contain any customer data, it's just metadata over global database accounts which contain the actual customer data.

sjanamma commented:
>                 "$ref": "#/definitions/ThroughputPoolCreateUpdateParameters"

PUT must have the same request and response models which should be same as GET

Path operation should have a different request model

---
In reply to: [1821843871](https://github.com/Azure/azure-rest-api-specs/pull/26206#issuecomment-1821843871) [](http://example.com/codeflow?ancestors=1821843871)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:162 in 90782ad. [](commit_id = 90782addcc11916078db97f65b1e36fc09ec4b04, deletion_comment = False)

sjanamma commented:
Please address all the review comments before pushing it to ARM Review queue

pjohari-ms committed:
Resolved review comments

pjohari-ms commented:
> Please address all the review comments before pushing it to ARM Review queue

The only remaining review comment was the one where I needed an explanation from the ARM reviewers, and hence sending it to the "Review" queue. All others were addressed.

I have addressed the remaining comment. I'm very confused on why we would want to have a GET request with an optional request body mentioned in the swagger, in ANY situation. I looked up "x-ms-mutability" documentation online https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-mutability and this makes sense mostly for response models only. Could you provide documentation/explanation on why we're using it for request models and why does it make sense to provide the same request body model to a GET request as well?

DrewKaufmann committed:
Added fields to support VPN (#26929)

* Update MI Cassandra preview swagger

* Fixed spellcheck and modelviolation checks

* Made model changes

* Changed param field names and added ListCommand op

* Made minor fixes

* Fixed ModelValidation checks

* Readded ARMProxyResource to BackupResource

* Removed ARMProxyResource from BackupResource

* Added fields to support VPN

pjohari-ms committed:
Merge branch 'main' into pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview

visingla-ms commented:
/azp run

visingla-ms commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1855 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/commands/{commandId}": {
+      "post": {
+        "operationId": "CassandraClusters_GetCommandAsync",
+        "description": "Get list of details about commands that were run asynchronously. Use \"all\" as commandId to list all commands in three days associated with the cluster.",
```
The description, response text etc. are all indicating it is for list of commands, whereas the URL is for commandId.

visingla-ms commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1855 @@
+          "description": "Result output of the command."
+        },
+        "status": {
+          "type": "string",
```
Enum?

visingla-ms commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -0,0 +1,2073 @@
+              "$ref": "#/definitions/OperationType"
+            },
+            "canUndelete": {
+              "type": "string",
```
Why is this string and not boolean here?

visingla-ms commented:
>                 "$ref": "#/definitions/ThroughputPoolGetResults"

Why is PUT request and response not matching?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:177 in a5315ab. [](commit_id = a5315abcfd7fb332c3d423180030e7ab17a2c280, deletion_comment = False)

visingla-ms commented:
>                 "$ref": "#/definitions/ThroughputPoolCreateUpdateParameters"

Why are there request payload parameters for GET?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:91 in a5315ab. [](commit_id = a5315abcfd7fb332c3d423180030e7ab17a2c280, deletion_comment = False)

visingla-ms commented:
>             "x-ms-long-running-operation": true,

For all long running PUT operations, the response should include 201 and 202, both.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:149 in a5315ab. [](commit_id = a5315abcfd7fb332c3d423180030e7ab17a2c280, deletion_comment = False)

visingla-ms commented:
>             }

The error schema is missing for these new operations.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:180 in a5315ab. [](commit_id = a5315abcfd7fb332c3d423180030e7ab17a2c280, deletion_comment = False)

visingla-ms commented:
>                         "$ref": "#/definitions/ThroughputPoolsListResult"

Please add pagination (x-ms-pageable).

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:58 in a5315ab. [](commit_id = a5315abcfd7fb332c3d423180030e7ab17a2c280, deletion_comment = False)

pjohari-ms committed:
Prettify ThroughpuutPool.json

pjohari-ms committed:
Resolved review comments

DrewKaufmann committed:
Resolved review comments (#27038)

* Resolved review comments

* Fixed errors in failed checks

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "Success. Return a list of command resource.",
```
same as above

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/commands/{commandId}": {
+      "get": {
+        "operationId": "CassandraClusters_GetCommandAsync",
+        "description": "Get list of details about commands that were run asynchronously. Use \"all\" for commandId to list all commands in three days associated with the cluster.",
```
why `description` use `list` for a single command?

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -791,6 +924,16 @@
       "maxLength": 100,
       "pattern": "^[a-zA-Z0-9]+(-[a-zA-Z0-9]+)*$"
     },
+    "commandId": {
```
please add `pattern:` for property in path

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -1345,6 +1538,81 @@
+        "arguments": {
+          "type": "object",
+          "description": "The arguments for the command to be run",
+          "additionalProperties": {
```
`additionalProperties` is not allowed. why not use a string array?

DrewKaufmann replied:
There are `additionalProperties` used elsewhere in other definition fields. Would those need to be changed in addition to this one to adhere to this standard?

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -1345,6 +1538,81 @@
+          "description": "Whether command has admin privileges"
+        },
+        "cassandraStopStart": {
+          "type": "boolean",
```
seems like to control different behaviors, please consider using enum instead:
https://github.com/Azure/azure-openapi-validator/blob/main/docs/enum-instead-of-boolean.md

DrewKaufmann replied:
I believe a boolean would work here, as it is either "true", meaning Cassandra requires it to stop, run command, and start again, or "false", meaning that a stop is not required to run command. There wouldn't need to be any additional actions besides the two.

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -1345,6 +1538,81 @@
+          "description": "If true, stops cassandra before executing the command and then start it again"
+        },
+        "readWrite": {
+          "type": "boolean",
```
same as above, consider using enum instead.

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1099,6 +1099,16 @@
               "description": "The operation type of this database event.",
               "$ref": "#/definitions/OperationType"
             },
+            "canUndelete": {
```
since the description says it's a `state`, maybe the name should be `xxxStatus`? And it should be an `enum`?

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1207,6 +1217,16 @@
               "description": "The operation type of this container event.",
               "$ref": "#/definitions/OperationType"
             },
+            "canUndelete": {
```
same as above

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1207,6 +1217,16 @@
+              "readOnly": true,
+              "description": "A state of this container to identify if this container is restorable in same account."
+            },
+            "canUndeleteReason": {
```
`unrestorableReason`?

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1351,6 +1371,16 @@
               "description": "The operation type of this database event.",
               "$ref": "#/definitions/OperationType"
             },
+            "canUndelete": {
```
same as above

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1351,6 +1371,16 @@
+              "readOnly": true,
+              "description": "A state of this database to identify if this database is restorable in same account."
+            },
+            "canUndeleteReason": {
```
same as above

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1430,6 +1460,16 @@
               "description": "The operation type of this collection event.",
               "$ref": "#/definitions/OperationType"
             },
+            "canUndelete": {
```
same as above

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/restorable.json:
```diff
@@ -1430,6 +1460,16 @@
+              "readOnly": true,
+              "description": "A state of this collection to identify if this container is restorable in same account."
+            },
+            "canUndeleteReason": {
```
same as above

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+        "responses": {
+          "202": {
+            "description": "Success. The command operation will begin as soon as possible.",
+            "schema": {
```
202 response for a LRO POST operation must not have a response schema specified.
https://github.com/Azure/azure-openapi-validator/blob/main/docs/post-response-codes.md

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
A 202 response should include an Location response header.
https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-location-header.md

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "cosmos-db.json#/definitions/CloudError"
```
Error response content of long running operations must follow the error schema provided in the common types v2 and above.
https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-error-content.md

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
`x-ms-pageable` extension must be specified for LIST APIs.
https://github.com/Azure/azure-openapi-validator/blob/main/docs/xms-pageable-for-list-calls.md

ms-zhenhua commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:
```diff
@@ -0,0 +1,587 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/throughputPools": {
```
please fix Lint Diff errors at `Location: Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json` in https://github.com/Azure/azure-rest-api-specs/pull/26206/checks?check_run_id=19566418083

DrewKaufmann committed:
Addressed review comments (#27107)

* Resolved review comments

* Fixed errors in failed checks

* Addressed review comments

* Fixed check violations

pjohari-ms committed:
LintDiff

guanzhousongmicrosoft committed:
Remove additionalProperty for MI Cassandra CommandPublicResource (#27122)

pjohari-ms committed:
More LintDiff errors

pjohari-ms committed:
Merge branch 'pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview' of https://github.com/Azure/azure-rest-api-specs into pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview

pjohari-ms committed:
More LintDiff errors

pjohari-ms committed:
LintDiff

guanzhousongmicrosoft committed:
Revert readWrite change for MI Cassandra (#27125)

* Remove additionalProperty for MI Cassandra CommandPublicResource

* Revert "readWrite" property to boolean

* fix case sensitive issue

* revert

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1915 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/commands/all": {
```
What's the difference between this API and above "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/commands"?

guanzhousongmicrosoft replied:
fixed

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -0,0 +1,1915 @@
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/cassandraClusters/{clusterName}/commands/all": {
+      "get": {
+        "operationId": "CassandraClusters_GetCommandAsync",
```
The GET operation should not be an async operation.

guanzhousongmicrosoft replied:
This operation is not async. This only means get the result immediately from an "async" dba command, which was introduced in this version.

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:
```diff
@@ -0,0 +1,675 @@
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/ThroughputPoolUpdateProperties"
```
It's not necessary to split the create/update properties, you could use the x-ms-mutability to differentiate the create/update.
refs: https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-mutability

ms-henglu commented:
Please also check the comments from previous review and make sure that you have addressed them all.

pjohari-ms committed:
Model validation errors

pjohari-ms committed:
Merge branch 'pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview' of https://github.com/Azure/azure-rest-api-specs into pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview

pjohari-ms committed:
Reviewer's comment about DataTransfer

guanzhousongmicrosoft committed:
Remove duplication (#27134)

pjohari-ms committed:
Spell check

pjohari-ms committed:
Merge branch 'pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview' of https://github.com/Azure/azure-rest-api-specs into pjohari-ms-cosmos-db-Microsoft.DocumentDB-2023-11-15-preview

pjohari-ms committed:
Prettier

pjohari-ms commented:
/azp run

pjohari-ms commented:
We have addressed all relevant review comments. Remaining LintDiff comments are for code that's not introduced in this PR, or is not relevant for us.

pjohari-ms commented:
Both spellCheck and prettier are pointing to non existent issues.

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/dataTransferService.json:
```diff
@@ -473,6 +469,23 @@
+          ],
+          "x-ms-enum": {
+            "name": "DataTransferJobMode",
+            "modelAsString": false
```
@pjohari-ms, please double check this comment.

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/managedCassandra.json:
```diff
@@ -341,6 +341,139 @@
+          "200": {
+            "description": "Success. Return a list of command resource.",
+            "schema": {
+              "$ref": "#/definitions/ListCommands"
```
@pjohari-ms, please reply this comment.

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:
```diff
@@ -0,0 +1,664 @@
+        "create",
+        "update"
+      ],
+      "properties": {
```
200 response schema in long running PUT operation is missing ProvisioningState property. A LRO PUT operations response schema must have ProvisioningState specified for the 200 and 201 status codes.

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:
```diff
@@ -0,0 +1,664 @@
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/ThroughputPoolResource"
```
Properties of a PATCH request body must not be required, property:location.

ms-henglu commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2023-11-15-preview/throughputpool.json:
```diff
@@ -0,0 +1,664 @@
+          }
+        }
+      },
+      "put": {
```
Synchronous and long-running PUT operations must have responses with 200, 201 and default return codes. They also must not have other response codes.

ms-henglu commented:
Hi @pjohari-ms , please run the `oav` locally to check the lint diff errors detected in the `throughput pool.json`, I believe this file is added in this PR.

refs: Synchronous and long-running PUT operations must have responses with 200, 201 and default return codes. They also must not have other response codes.

example: `autorest --v3 --spectral --azure-validator --use=@microsoft.azure/openapi-validator@2.1.6 --tag=package-pre
view-2023-11 /Users/xxx/go/src/github.com/Azure/azure-rest-api-specs/specification/cosmos-db/resource-manager/readme.md`

pjohari-ms commented:
@ms-henglu I will fix the issues you pointed out in next iteration. Are there any other issues remaining? Or the PR should be good to merge after that?

pjohari-ms commented:
Asking this because oav lintdiff points to a lot of issues which might not be relevant.

ms-henglu commented:
> Asking this because oav lintdiff points to a lot of issues which might not be relevant.

I think once the errors detected in  throughput pool.json are fixed, we're good to go.

pjohari-ms committed:
Corrected Spec

pjohari-ms committed:
More validation errors

pjohari-ms committed:
Prettier nit

pjohari-ms commented:
@ms-henglu 
1. All model validation errors fixed
2. No lintDiff errors present for throughputPool.json
3. Rest others are for different resources not introduced in this PR.

Please take a look and merge if it meets all the criteria. Thanks.

ms-henglu commented:
Hi @pjohari-ms , in order to get this PR merged, please get approval for the following sdk breaking changes:
❌ Approved-SdkBreakingChange-Go
❌ Approved-SdkBreakingChange-Python
❌ Approved-SdkBreakingChange-JavaScript

pjohari-ms commented:
@msyyc @Alancere @kazrael2119 could you take a look at the SDK breaking changes in this PR? These changes related to managedCassandra are added to fix changes introduced in the previous 09-15-preview version. Since this is a preview version, we should be fine to release these changes to our customers anyway.
cc @guanzhousongmicrosoft @DrewKaufmann

msyyc commented:
Approved for Python

pjohari-ms commented:
/pr RequestMerge

ms-henglu commented:
/azp run

ms-henglu commented:
Hi @pjohari-ms , the cspell pipeline has errors that blocked the merge, would you please check whether it's related with your "cspell.json" changes?

pjohari-ms committed:
Nit related to cSpell. Guessing game since local validation is not present.

pjohari-ms committed:
One more nit

pjohari-ms commented:
@ms-henglu Fixed the issue. Please merge.
